libsolidity/formal/SMTChecker.cpp

```diff
@@ -90,42 +91,66 @@ bool SMTChecker::visit(IfStatement const& _node)
 
 	checkBooleanNotConstant(_node.condition(), "Condition is always $VALUE.");
 
-	visitBranch(_node.trueStatement(), expr(_node.condition()));
+	auto countersEndFalse = m_currentSequenceCounter;
+	auto countersEndTrue = visitBranch(_node.trueStatement(), expr(_node.condition()));
+
 	vector<Declaration const*> touchedVariables = m_variableUsage->touchedVariables(_node.trueStatement());
 	if (_node.falseStatement())
 	{
-		visitBranch(*_node.falseStatement(), !expr(_node.condition()));
+		countersEndFalse = visitBranch(*_node.falseStatement(), !expr(_node.condition()));
 		touchedVariables += m_variableUsage->touchedVariables(*_node.falseStatement());
 	}
 
-	resetVariables(touchedVariables);
+	mergeVariables(touchedVariables, expr(_node.condition()), countersEndTrue, countersEndFalse);
 
 	return false;
 }
 
 bool SMTChecker::visit(WhileStatement const& _node)
 {
-	auto touchedVariables = m_variableUsage->touchedVariables(_node);
-	resetVariables(touchedVariables);
+	auto countersEndFalse = m_currentSequenceCounter;
+	if (m_outterLoop == nullptr)
+		m_outterLoop = &_node;
 	if (_node.isDoWhile())
 	{
 		visitBranch(_node.body());
 		// TODO the assertions generated in the body should still be active in the condition
+		m_currentLoopCondition = &_node.condition();
 		_node.condition().accept(*this);
-		checkBooleanNotConstant(_node.condition(), "Do-while loop condition is always $VALUE.");
+		//checkBooleanNotConstant(_node.condition(), "Do-while loop condition is always $VALUE.");
+		m_currentLoopCondition = nullptr;
 	}
 	else
 	{
+		m_currentLoopCondition = &_node.condition();
 		_node.condition().accept(*this);
-		checkBooleanNotConstant(_node.condition(), "While loop condition is always $VALUE.");
-
-		visitBranch(_node.body(), expr(_node.condition()));
+		if (m_maximizationResult.first == smt::CheckResult::UNSATISFIABLE)
+			m_errorReporter.warning(
+				_node.location(),
+				"While loop condition is always false."
+			);
+		m_currentLoopCondition = nullptr;
+	}
+	decltype(countersEndFalse) countersEndTrue;
+	for(int i = 0; i < computeLoopBound(); ++i)
+	{
+		_node.condition().accept(*this);
```

Comment:
Is it really fine to visit it again?

Reply:
I had to adjust expression creation for that, by using the new `m_outterLoop`. The idea is that if we're currently visiting the condition or body of a loop we allow expressions to be created again, just with an incremented index in the name as well. Because of that, every time the condition is visited, it actually uses the updated variables (which are actually just the normal `ite` merged vars).